File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let answ = []
    backtrack([], new Set(nums))
    function backtrack(current, left){
        if(current.length == nums.length){
            answ.push([...current])
            return
        }

        for(let i = 0; i < nums.length; i++){
            if(left.has(nums[i])){
                left.delete(nums[i])
                backtrack([...current, nums[i]], left)
                left.add(nums[i])
            }
        }
    }
    return answ
};

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> answ;
        vector<int> soFar;
        permuteClone(soFar, nums, answ);
        return answ;
    }
    void permuteClone(vector<int>& soFar, vector<int>& nums, vector<vector<int>>& answ){
        if(nums.empty()){
            answ.push_back(soFar);
            return;
        }
        for(int i = 0; i < nums.size(); i++){
            int currNum = nums[i];
            nums.erase(nums.begin() + i);
            soFar.push_back(currNum);
            permuteClone(soFar, nums, answ);
            nums.insert(nums.begin() + i, currNum);
            soFar.pop_back();
        }
    }
};